sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -152,3 +152,99 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
+    if (SQLConf.get.constraintPropagationEnabled) {
+      eliminateCrossJoin(plan)
+    } else {
+      plan
+    }
+  }
+
+  private def eliminateCrossJoin(plan: LogicalPlan): LogicalPlan = plan transform {
+    case join@Join(leftPlan, rightPlan, Cross, None) =>
+      val leftConstraints = join.constraints.filter(_.references.subsetOf(leftPlan.outputSet))
+      val rightConstraints = join.constraints.filter(_.references.subsetOf(rightPlan.outputSet))
+      val inferredJoinPredicates = inferJoinPredicates(leftConstraints, rightConstraints)
+      val joinConditionOpt = inferredJoinPredicates.reduceOption(And)
+      if (joinConditionOpt.isDefined) Join(leftPlan, rightPlan, Inner, joinConditionOpt) else join
+  }
+
+  private def inferJoinPredicates(
+      leftConstraints: Set[Expression],
+      rightConstraints: Set[Expression]): Set[EqualTo] = {
+
+    // iterate through the left constraints and build a hash map that points semantically
+    // equivalent expressions into attributes
+    val emptyEquivalenceMap = Map.empty[SemanticExpression, Set[Attribute]]
+    val equivalenceMap = leftConstraints.foldLeft(emptyEquivalenceMap) { case (map, constraint) =>
+      constraint match {
+        case EqualTo(attr: Attribute, expr: Expression) =>
+          updateEquivalenceMap(map, attr, expr)
+        case EqualTo(expr: Expression, attr: Attribute) =>
+          updateEquivalenceMap(map, attr, expr)
+        case _ => map
+      }
+    }
+
+    // iterate through the right constraints and infer join conditions using the equivalence map
+    rightConstraints.foldLeft(Set.empty[EqualTo]) { case (joinConditions, constraint) =>
+      constraint match {
+        case EqualTo(attr: Attribute, expr: Expression) =>
+          appendJoinConditions(attr, expr, equivalenceMap, joinConditions)
+        case EqualTo(expr: Expression, attr: Attribute) =>
+          appendJoinConditions(attr, expr, equivalenceMap, joinConditions)
+        case _ => joinConditions
+      }
+    }
+  }
+
+  private def updateEquivalenceMap(
+      equivalenceMap: Map[SemanticExpression, Set[Attribute]],
+      attr: Attribute,
+      expr: Expression): Map[SemanticExpression, Set[Attribute]] = {
+
+    val equivalentAttrs = equivalenceMap.getOrElse(expr, Set.empty[Attribute])
+    if (equivalentAttrs.contains(attr)) {
+      equivalenceMap
+    } else {
+      equivalenceMap.updated(expr, equivalentAttrs + attr)
+    }
+  }
+
+  private def appendJoinConditions(
+      attr: Attribute,
+      expr: Expression,
+      equivalenceMap: Map[SemanticExpression, Set[Attribute]],
+      joinConditions: Set[EqualTo]): Set[EqualTo] = {
+
+    equivalenceMap.get(expr) match {
+      case Some(equivalentAttrs) => joinConditions ++ equivalentAttrs.map(EqualTo(attr, _))
+      case None => joinConditions
+    }
+  }
+
+  // the purpose of this class is to treat 'a === 1 and 1 === 'a as the same expressions
+  implicit class SemanticExpression(private val expr: Expression) {
```

Comment:
Can we reuse `EquivalentExpressions`? You can search the code base and see how the others use it.

Reply:
@gatorsmile 

I think we just need the case class inside ``EquivalentExpressions`` since we have to map all semantically equivalent expressions into a set of attributes (as opposed to mapping an expression into a set of equivalent expressions). 

I see two ways to go:

1. Expose the case class inside ``EquivalentExpressions`` with minimum changes in the code base (e.g., using a companion object):

````
object EquivalentExpressions {

  /**
   * Wrapper around an Expression that provides semantic equality.
   */
  implicit class SemanticExpr(private val e: Expression) {
    override def equals(o: Any): Boolean = o match {
      case other: SemanticExpr => e.semanticEquals(other.e)
      case _ => false
    }

    override def hashCode: Int = e.semanticHash()
  }
}
````

2. Keep ``EquivalentExpressions`` as it is and maintain a separate map from expressions to attributes in the proposed rule.

Personally, I lean toward the first idea since it might be useful to have ``SemanticExpr`` alone. However, there can be other drawbacks that did not come into my mind.

Reply:
How about building a new class to process all the cases similar to this one?

An `Attribute` is also an `Expression`. Basically, the internal will be still a hash map `mutable.HashMap.empty[SemanticEqualExpr, mutable.MutableList[Expression]]`

Reply:
Using a Set instead of a List might be beneficial in the proposed rule. What about the following?

```
class EquivalentExpressionMap {

  private val equivalenceMap = mutable.HashMap.empty[SemanticallyEqualExpr, mutable.Set[Expression]]

  def put(expression: Expression, equivalentExpression: Expression): Unit = {
    val equivalentExpressions = equivalenceMap.getOrElse(expression, mutable.Set.empty)
    if (!equivalentExpressions.contains(equivalentExpression)) {
      equivalenceMap(expression) = equivalentExpressions += equivalentExpression
    }
  }
  
  // produce an immutable copy to avoid any modifications from outside
  def get(expression: Expression): Set[Expression] =
    equivalenceMap.get(expression).fold(Set.empty[Expression])(_.toSet)

}

object EquivalentExpressionMap {

  private implicit class SemanticallyEqualExpr(private val expr: Expression) {
    override def equals(o: Any): Boolean = o match {
      case other: SemanticallyEqualExpr => expr.semanticEquals(other.expr)
      case _ => false
    }

    override def hashCode: Int = expr.semanticHash()
  }
}

```

Reply:
I did not check it carefully, but how about `ExpressionSet`?

Reply:
I am afraid ``ExpressionSet`` will not help here since we need to map a semantically equivalent expression into a set of attributes that correspond to it. It is not enough to check if there is an equivalent expression. Therefore, ``EquivalentExpressions`` and ``ExpressionSet`` are not applicable (as far as I see). 

``EquivalentExpressionMap`` from the previous comment assumes the following workflow:

```
val equivalentExressionMap = new EquivalentExpressionMap
...
equivalentExressionMap.put(1 * 2, t1.a)
equivalentExressionMap.put(3, t1.b)
...
equivalentExressionMap.get(1 * 2) // Set(t1.a)
equivalentExressionMap.get(2 * 1) // Set(t1.a)

```

Reply:
I mean `def get(expression: Expression): ExpressionSet`